### PR TITLE
Improve bad redirect query

### PR DESCRIPTION
**Pull request by sauyon**
Fixes false positives by creating a barrier guard with the bad check, then checking for flow from a node guarded by that barrier guard into a redirect call.

Also removes the heuristic part of the query, which was causing a false negative.

Currently depends on #110 and #112, which are required to detect the CVEs.

### CODE REVIEW

File: ql/src/Security/CWE-601/BadRedirectCheck.ql [hunk 1]
```diff
@@ -12,52 +12,112 @@
...
   )
 }
 
-from DataFlow::Node node, SsaWithFields v
+class Configuration extends TaintTracking::Configuration {
+  Configuration() { this = "BadRedirectCheck" }
+
+  override predicate isSource(DataFlow::Node source) { this.isSource(source, _) }
+
+  predicate isSource(DataFlow::Node source, BadRedirectCheckBarrier check) {
```
Comment (max-schaefer) [inline ql/src/Security/CWE-601/BadRedirectCheck.ql:64]: Needs qldoc.

File: ql/src/Security/CWE-601/BadRedirectCheck.ql [hunk 2]
```diff
@@ -12,52 +12,112 @@
...
     slash.getStringValue() = "\\" and
-    er.getBase() = v.getAUse() and
+    er.getBase() = checked and
     er.getIndex().getIntValue() = 1 and
     eq.eq(_, er, slash)
   )
 }
 
-from DataFlow::Node node, SsaWithFields v
+class Configuration extends TaintTracking::Configuration {
```
Comment (max-schaefer) [inline ql/src/Security/CWE-601/BadRedirectCheck.ql:59]: Needs qldoc.

File: ql/src/Security/CWE-601/BadRedirectCheck.ql [hunk 3]
```diff
@@ -12,52 +12,112 @@
...
+  Configuration() { this = "BadRedirectCheck" }
+
+  override predicate isSource(DataFlow::Node source) { this.isSource(source, _) }
+
+  predicate isSource(DataFlow::Node source, BadRedirectCheckBarrier check) {
+    source = check.getAGuardedNode()
+    or
+    exists(DataFlow::Node pred | pred.getASuccessor() = source | check.guardsFlowFrom(pred))
```
Comment (max-schaefer) [inline ql/src/Security/CWE-601/BadRedirectCheck.ql:67]:
```suggestion
    check.guardsFlowFrom(source.getAPredecessor())
```

File: ql/src/Security/CWE-601/BadRedirectCheck.ql [hunk 4]
```diff
@@ -12,52 +12,112 @@
...
+    or
+    exists(
+      DeclaredFunction f, FunctionInput inp, FunctionOutput outp,
+      ControlFlow::ConditionGuardNode guard, BadRedirectCheckBarrier check
+    |
+      check.checks(inp.getExitNode(f.getFuncDecl()).getASuccessor*().asExpr(), _) and
+      guard.ensures(outp.getEntryNode(f.getFuncDecl()), _) and
+      TaintTracking::localTaint(check, outp.getEntryNode(f.getFuncDecl())) and
+      this = f.getACall() and
+      checked = inp.getNode(this).asExpr()
```
Comment (max-schaefer) [inline ql/src/Security/CWE-601/BadRedirectCheck.ql:110]: This looks very similar to the logic added to `guardingFunction` below. Is there any way to share the logic?

**Comment by sauyon** [comment]
That probably means this requires more comments.

**Comment by sauyon** [comment]
Closing c.f. the discussion in #115.

**Pull request closed without merging**